### PR TITLE
issue: 897296 Fix advanced routing issues

### PR DESCRIPTION
**Pull request by ashanably**
- Let connect to double resolve route distination when source address isn't assigned

- Resolve the correct outgoing interface when the interface that hold route entry source address  is different than route entry interface index

Signed-off-by: ashanably <ashanably@asaltech.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/609/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/610/ for details (Mellanox internal link).

File: src/vma/proto/dst_entry.cpp [hunk 1]
```diff
@@ -70,7 +70,7 @@ dst_entry::~dst_entry()
 	}
 
 	if (m_p_rt_entry) {
-		g_p_route_table_mgr->unregister_observer(route_rule_table_key(m_dst_ip.get_in_addr(), m_bound_ip ? m_bound_ip : m_so_bindtodevice_ip, m_tos), this);
+		g_p_route_table_mgr->unregister_observer(route_rule_table_key(m_dst_ip.get_in_addr(), m_src_ip, m_tos), this);
```
Comment (ilansmith) [inline src/vma/proto/dst_entry.cpp]: Lets go over the changing of src_ip selection F2F - in particular the changing of m_bound_ip an m_so_bindtodevice_ip usage.

File: src/vma/proto/dst_entry.h [hunk 2]
```diff
@@ -65,7 +65,7 @@ class dst_entry : public cache_observer, public tostr, public neigh_observer
 
 	virtual void 	notify_cb();
 
-	virtual bool 	prepare_to_send(bool skip_rules=false);
+	virtual bool 	prepare_to_send(bool skip_rules=false, bool is_connect=false);
```
Comment (ilansmith) [inline src/vma/proto/dst_entry.h]: I've revised this again.
In principal `is_connect` shouldn't be exposed to `dst_entry_udp` and `dst_entry_udp_mc` just as neither it nor `skip_rules` should be exposed to UDP in `prepare_to_send()`. In all these cases UDP always passes `false`.
The inheritance of `prepare_to_send()` and `resolve_net_dev()` should be fixed up together in a separate commit.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/615/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/611/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/646/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/640/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/647/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/641/ for details (Mellanox internal link).

File: src/vma/proto/dst_entry.cpp [hunk 3]
```diff
@@ -251,6 +265,9 @@ bool dst_entry::resolve_net_dev(bool is_connect)
 
 	if (update_rt_val()) {
 		ret_val = update_net_dev_val();
+		if (ret_val) {
+			set_src_addr();
```
Comment (ilansmith) [inline src/vma/proto/dst_entry.cpp]: Why not do this from within update_net_dev_val()?

File: src/vma/proto/dst_entry.h [hunk 4]
```diff
@@ -77,15 +77,7 @@ class dst_entry : public cache_observer, public tostr, public neigh_observer
 	in_addr_t	get_dst_addr();
 	uint16_t	get_dst_port();
 	inline in_addr_t get_src_addr() const {
```
Comment (ilansmith) [inline src/vma/proto/dst_entry.h]: I don't see why this should be any different than the functions above (get_dst_addr/get_dst_port) with respect to being inline and having the implementation here rather than in the cpp file.

File: src/vma/proto/dst_entry.cpp (hunk 3 above)
Comment (ashanably) [inline src/vma/proto/dst_entry.cpp]: In face setting source address is not directly related to net_dev_val and to move it to update_net_dev_val it will appear that it depend on this function given that it depend route resolving and update_rt_val() as well.

File: src/vma/proto/dst_entry.h (hunk 4 above)
Comment (ashanably) [inline src/vma/proto/dst_entry.h]: We agreed in meeting to keep it inline and that what is done. In general this function was not inlined before pull request changes so we can return it back.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/653/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/646/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
Merged #242
